velox/common/memory/MmapAllocator.h

```diff
@@ -0,0 +1,226 @@
+        MappedMemory::Allocation* out);
+
+    // Frees all pages of 'allocation' that fall in this size
+    // class. Erases the corresponding runs from 'allocation'.
+    MachinePageCount free(Allocation* allocation);
+
+    ClassPageCount checkConsistency(ClassPageCount* numMapped);
+
+    MachinePageCount adviseAway(
+        MachinePageCount numPages,
+        MmapAllocator* allocator);
+
+    void setAllMapped(Allocation* allocation, bool value);
+    void setMappedBits(MappedMemory::PageRun run, bool value);
+    bool isAllocated(uint8_t* ptr) {
+      if (ptr >= address_ && ptr < address_ + capacity_ * pageSize_) {
+        uint64_t offset = (ptr - address_) / pageSize_;
+        return (pageAllocated_[offset / 64] & (1L << offset & 63)) != 0;
+      }
+      return false;
+    }
+    bool isInRange(uint8_t* ptr) {
+      if (address_ > ptr || address_ + byteSize_ <= ptr) {
+        return false;
+      }
+      // See that ptr falls on a page boundary.
+      if ((ptr - address_) % pageSize_ != 0) {
+        VELOX_FAIL("Pointer is in a SizeClass but not at page boundary");
+      }
+      return true;
+    }
+    std::string toString() const;
+
+   private:
+    bool allocateLocked(
+        ClassPageCount numPages,
+        int owner,
+        MachinePageCount* numUnmapped,
+        MappedMemory::Allocation* out);
+
+    void adviseAway(Allocation* allocation);
+
+    void allocMapped(
+        int wordIndex,
+        uint64_t candidates,
+        ClassPageCount* needed,
+        MappedMemory::Allocation* allocation);
+
+    void allocAny(
+        int wordIndex,
+        ClassPageCount* needed,
+        MachinePageCount* numUnmapped,
+        Allocation* allocation);
+
+    size_t capacity_;
+    MachinePageCount pageSize_;
+    uint8_t* address_;
+    size_t byteSize_;
+    int clockHand_ = 0;
+    // Count of free pages backed by memory.
+    ClassPageCount numMappedFreePages_ = 0;
+    // 1 bit per page.
+    std::vector<uint64_t> pageAllocated_;
+    std::vector<uint64_t> pageMapped_;
+    std::vector<int> pageOwner_;
+
+    // Statistics.
+    uint64_t numAllocatedMapped_ = 0;
+    uint64_t numAllocatedUnmapped_ = 0;
+    uint64_t numAdvisedAway_ = 0;
+    std::mutex mutex_;
+  };
+
+  // Marks all the pages backing 'out' to be mapped if one can safely
+  // write up to 'numMappedNeeded' pages that have no backing
+  // memory. Returns true on success. Returns false if enough backed
+  // but unused pages from other size classes cannot be advised
+  // away. On failure, also frees 'out'. since the whole allocation
+  // could not be completed.
+  bool ensureEnoughMappedPages(int32_t newMappedNeeded, Allocation& out);
+
+  // Frees 'allocation and returns the number of freed pages. Does not
+  // update 'numAllocated'.
+  MachinePageCount freeInternal(Allocation& allocation);
+
+  // Returns the index of the size class 'address' falls in. Throws if 'address'
+  // is not in any.
+  int sizeClass(uint8_t* address);
```

Comment:
- make this method const ?
- inconsistent naming: address vs ptr in other methods